src/bot.rs

```diff
@@ -0,0 +1,175 @@
+    game_last_played: &mut HashMap<String, std::time::Instant>,
+    client: &mut Telegram
+) {
+    use telegram_bot_raw::types::update::UpdateKind;
+    use telegram_bot_raw::types::message::MessageKind;
+    if let UpdateKind::Message(msg) = update.kind {
+        if let MessageKind::Text{data, entities} = msg.kind {
+            drop(entities); // Silence the stupid warning and free some RAM
+            if data.contains("/start") {
+                let pieces: Vec<String> = data.split(" ").map(|x| x.to_owned()).collect();
+                if pieces.len() == 1 {
+                    client.send_message(("Ciao! A che gioco vuoi giocare?", msg.from.id, playable_games).into());
+                } else {
+                    let game_id = pieces[1].clone();
+                    add_player_to_game(game_id, client, player_games, game_channel, msg.from);
+                }
+            } else {
+                // Pass to thread
+                // It's a text message that has to be handled. If a user has more than one active game
+                // I have to ask him which one
+            }
+        } // ignoring other message kinds since they're useless for us
+    }  else if let UpdateKind::CallbackQuery(qry) = update.kind {
+        handle_callback_query(qry, playable_games, player_games, game_channel, game_last_played, client);
+    }
+}
+
+pub fn main_bot_logic(
+    playable_games: &Vec<Box<dyn Game>>,
+    player_games: &mut HashMap<telegram_bot_raw::types::refs::UserId, String>,
+    game_channel: &mut HashMap<String, std::sync::mpsc::SyncSender<threading::ThreadMessage>>,
+    game_last_played: &mut HashMap<String, std::time::Instant>,
+    client: &mut Telegram
+) {
+    let mut cleanup_list: Vec<String> = Vec::new();
+    loop {
+        for update in client.get_updates() {
+            handle_update(update, playable_games, player_games, game_channel, game_last_played, client);
+        }
+        // Now check if there are games which need to be terminated
+        for (game, time) in game_last_played.iter() {
+            if time.elapsed().as_secs() > MAX_GAME_DURATION {
+                let channel = game_channel.get(game).unwrap();
+                channel.send(threading::ThreadMessage::Kill).unwrap_or_default();
+            } else if time.elapsed().as_secs() > (MAX_GAME_DURATION as f64 * 0.9) as u64 {
+                let channel = game_channel.get(game).unwrap();
+                channel.send(threading::ThreadMessage::AboutToKill).unwrap_or_default();
+            }
+        }
+        // Cleanup of dead games
+        for (game, channel) in game_channel.iter() {
+            if channel.send(threading::ThreadMessage::Ping).is_err() {
+                // If the game is dead, the channel can't send the message
+                cleanup_list.push(game.clone());
+            }
+        }
+        // Deassosciate the game
+        *player_games = player_games.iter()
```

Comment:
this has to change consistently with previous comments

Reply:
Yes, maybe a `cleanup_old_games` function could be a solution